### PR TITLE
Support unicode superscripts

### PR DESCRIPTION
**Pull request by yongqli**
Currently Rust allows unicode subscripts to continue identifiers; I've added support for superscripts as well.

See below for discussions:
https://github.com/rust-lang/rust/issues/4928
https://users.rust-lang.org/t/allow-unicode-superscripts-in-identifiers/1892

r? @nikomatsakis

### CODE REVIEW

File: src/test/run-pass/unicode-superscripts.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    let σ⁰ = 2;
+    let σ¹ = 2;
+    let σ² = 1;
+    let σ³ = 2;
+    let σ⁴ = 2;
+    let σ⁵ = 2;
+    let σ⁶ = 2;
+    let σ⁷ = 2;
+    let σ⁸ = 2;
+    let σ⁹ = 2;
```
Comment (killercup) [inline src/test/run-pass/unicode-superscripts.rs]: Maybe also test `let σ⁴² = 2;`

**Comment by cmr** [comment]
I'm not in favor. Get the Unicode Consortium to extend XID_continue, let's not deviate from them.

**Comment by cmr** [comment]
(Note also that I believe extending the surface syntax in this way would require an RFC.)

**Comment by petrochenkov** [comment]
-1 to this random extension.
#4928 can be resolved by moving from Default Identifier Syntax to richer Alternative Identifier Syntax (with amendments), as defined by [UAX #31](http://unicode.org/reports/tr31/) (this is what C++ did, so Clang currently accept `σ⁴²` as an identifier).
The paper linked from #4928 contains a good overview of the problem and related standards.

**Comment by steveklabnik** [comment]
> (Note also that I believe extending the surface syntax in this way would require an RFC.)

Given that unicode identifiers are still behind a feature flag, I don't think so. There would be one to stabilize it, of course...

**Comment by nagisa** [comment]
Alternatively, we could accept all characters from Lm category as non-first identifier character. This is what [Haskell] did.

[Haskell]: https://ghc.haskell.org/trac/ghc/ticket/10196

**Comment by nagisa** [comment]
Given the controversy of this, this should probably be an [rfc](https://github.com/rust-lang/rfcs).

**Comment by nikomatsakis** [comment]
Hmm. This has historically been an area with a (somewhat) surprising amount of controversy. I am not persuaded that we have to block on the Unicode consortium, but it is nice to follow existing precedent. Have other languages deviated from `XID_Continue`?

**Comment by nikomatsakis** [comment]
Many thanks to the author of this PR, but I am inclined not to accept it. I think instead someone who cares should write-up an RFC describing a coherent policy, which we should then implement and stabilize. Speaking personally, I think that an RFC that can argue from some precedent (C++, Haskell, etc) will be much stronger than what which carves out a new space of identifiers, regardless of how good that space is (I am unclear whether this PR is carving out a new space or not, so I'm not implying anything about this PR per se).

**Comment by nikomatsakis** [comment]
FWIW -- I'd love to be able to use unicode superscripts :)

**Pull request closed without merging**